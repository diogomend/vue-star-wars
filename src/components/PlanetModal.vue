<template>
  <v-dialog
    max-width="290"
    persistent
    v-model="dialog"
    @click:outside="hidePlanetModal"
  >
    <v-card light>
      <template v-if="!activePlanet">
        <div class="text-center">
          <v-progress-circular
            :size="50"
            color="primary"
            indeterminate
            class="mt-10 mb-10"
          />
        </div>
      </template>
      <template v-else>
        <v-card-title>{{ activePlanet.name }}</v-card-title>
        <v-card-text>
          <v-row v-for="(property, index) in properties" :key="index">
            <v-col cols="12" sm="5" class="modal-heading">{{
              property.name
            }}</v-col>
            <v-col cols="12" sm="7">
              {{ property.value }}
            </v-col>
          </v-row>
        </v-card-text>
      </template>
    </v-card>
  </v-dialog>
</template>

<script>
import { mapGetters, mapActions } from "vuex";
export default {
  methods: {
    ...mapActions("planets", ["hidePlanetModal"])
  },
  computed: {
    ...mapGetters({
      activePlanet: ["planets/getActivePlanet"],
      showPlanetModal: ["planets/showPlanetModal"]
    }),
    properties() {
      if (this.activePlanet) {
        return [
          { name: "Diameter", value: this.activePlanet.diameter },
          { name: "Climate", value: this.activePlanet.climate },
          { name: "Population", value: this.activePlanet.population }
        ];
      }

      return [];
    },
    isPlanetSelected() {
      return !!this.activePlanet;
    },
    dialog: {
      set() {},
      get() {
        return this.showPlanetModal;
      }
    }
  }
};
</script>

<style>
.modal-heading {
  font-weight: bold;
  font-size: 14px;
}

::v-deep .loading-modal {
  height: 200px;
}
</style>
