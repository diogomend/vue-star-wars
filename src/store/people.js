import swapi from "@/apis/swapi";
import storageHelper from "@/helpers/localStorage";

const actions = {
  fetchPeople: async ({ commit }) => {
    const storedPeople = storageHelper.getFromStorage();

    if (storedPeople && storedPeople.length) {
      commit("SET_PEOPLE", JSON.parse(storedPeople));
      return;
    }
    await getAllPeople(commit);
  }
};

const mutations = {
  SET_PEOPLE(state, people) {
    state.people = people;
  }
};

const getters = {
  getPeople: state => state.people
};

const state = () => ({
  people: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};

export const getAllPeople = async commit => {
  let currentPage = 1;
  let hasNextPage = true;
  let people = [];

  while (hasNextPage) {
    const { data } = await swapi.get("people", {
      params: {
        page: currentPage
      },
      useCache: true
    });

    people = people.concat(data.results);
    storageHelper.setItem(JSON.stringify(people));
    commit("SET_PEOPLE", people);

    hasNextPage = !!data.next;
    currentPage += 1;
  }

  return people;
};
