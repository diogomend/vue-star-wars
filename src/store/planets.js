import swapi from "@/apis/swapi";

const actions = {
  showPlanetDetails: async ({ commit }, planetID) => {
    commit("SET_SHOW_PLANET_MODAL", true);
    const { data } = await swapi.get(`planets/${planetID}`);
    commit("SET_SELECTED_PLANET", data);
  },
  hidePlanetModal: async ({ commit }) => {
    commit("SET_SHOW_PLANET_MODAL", false);
  }
};
const mutations = {
  SET_SELECTED_PLANET(state, planet) {
    state.selectedPlanet = planet;
  },
  SET_SHOW_PLANET_MODAL(state, show) {
    state.showPlanetModal = show;
  }
};

const getters = {
  getActivePlanet: state => state.selectedPlanet,
  showPlanetModal: state => state.showPlanetModal
};

const state = () => ({
  selectedPlanet: null,
  showPlanetModal: false
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
