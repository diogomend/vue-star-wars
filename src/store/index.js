import Vue from "vue";
import Vuex from "vuex";
import people from "@/store/people";
import planets from "@/store/planets";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    people,
    planets
  }
};

export default new Vuex.Store(storeOptions);
