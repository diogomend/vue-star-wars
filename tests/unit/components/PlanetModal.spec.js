import Vuetify from '@/plugins/vuetify';
import Vue from 'vue';
import Vuex from "vuex";
import PlanetModal from "@/components/PlanetModal.vue";
import { shallowMount } from "@vue/test-utils"
import { storeOptions } from "@/store/index"
Vue.use(Vuex);
Vue.use(Vuetify);

describe("PlanetModal.vue", () => {
    let store;

    beforeEach(() => {
        jest.clearAllMocks();
        store = new Vuex.Store(storeOptions);
      });

      it("is a Vue instance", () => {
        const wrapper = shallowMount(PlanetModal, {
          store
        });
    
        const template = wrapper.html();
        expect(template).toMatchSnapshot();
      });
      it("Properties to have info if activePlanet", () => {
        storeOptions.modules.planets.getters.getActivePlanet = jest.fn().mockImplementation(() => {
            return {diameter: 1, population: 2, climate: 'arid'}
        });
        store = new Vuex.Store(storeOptions);
        const wrapper = shallowMount(PlanetModal, {
          store
        });
        wrapper.vm.dialog = true;
    
        expect(wrapper.vm.properties.length).toBe(3);
        expect(wrapper.vm.isPlanetSelected).toBeTruthy();
      });

      it("Properties should be empty if no activePlanet", () => {
        storeOptions.modules.planets.getters.getActivePlanet = jest.fn().mockImplementation(() => {
            return false
        });
        store = new Vuex.Store(storeOptions);
        const wrapper = shallowMount(PlanetModal, {
          store
        });
    
        expect(wrapper.vm.properties.length).toBe(0);
        expect(wrapper.vm.isPlanetSelected).toBeFalsy();
      });
})