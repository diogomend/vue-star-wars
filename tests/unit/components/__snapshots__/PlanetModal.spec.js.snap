// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanetModal.vue is a Vue instance 1`] = `
<v-dialog max-width="290" persistent="">
  <v-card light="">
    <div class="text-center">
      <v-progress-circular size="50" color="primary" indeterminate="" class="mt-10 mb-10"></v-progress-circular>
    </div>
  </v-card>
</v-dialog>
`;
